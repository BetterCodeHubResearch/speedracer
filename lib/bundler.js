// Native
const path = require('path')

// Packages
const commonjs = require('rollup-plugin-commonjs')
const resolve = require('rollup-plugin-node-resolve')
const rollupStream = require('rollup-stream')

const createBundleStream = filename =>
rollupStream({
  entry: filename,
  format: 'iife',
  moduleName: 'run',
  plugins: [
    {
      resolveId(id) {
        if (id === 'speedracer') {
          return path.resolve(__dirname, './runner-client.js')
        }
      }
    },
    resolve({ jsnext: true, main: true }),
    commonjs()
  ]
})

module.exports = createBundleStream
