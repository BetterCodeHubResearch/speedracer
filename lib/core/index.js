export { default as Race } from './race'
export { default as Serie } from './serie'
