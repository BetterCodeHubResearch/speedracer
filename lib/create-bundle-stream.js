// Native
const path = require('path')

// Packages
const rollupStream = require('rollup-stream')
const resolve = require('rollup-plugin-node-resolve')
const commonjs = require('rollup-plugin-commonjs')

const createBundleStream = filename =>
  rollupStream({
    entry: filename,
    format: 'iife',
    moduleName: 'run',
    plugins: [
      {
        resolveId(id) {
          if (id === 'speedracer') {
            return path.resolve(__dirname, './runtime.js')
          }
        }
      },
      resolve({ jsnext: true, main: true }),
      commonjs()
    ]
  })

module.exports = createBundleStream
