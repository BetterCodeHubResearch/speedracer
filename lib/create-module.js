module.exports = (module, options) => {
  const createModule = (module, ...deps) => {
    return Promise.all(deps.map(dep => (
      Promise.resolve().then(() => (
        Array.isArray(dep) ? createModule(...dep) : createModule(dep)
      ))
    )))
    .then(deps => {
      console.log(module)
      return Promise.resolve().then(() => (
        Array.isArray(deps) ? module(...deps, options) : module(options)
      ))
    })
  }

  return createModule(
    Array.isArray(module) ? createModule(...module) : createModule(module)
  )
}
