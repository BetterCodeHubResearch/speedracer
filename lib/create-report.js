// Native
const path = require('path')

// Packages
const DevtoolsTimelineModel = require('devtools-timeline-model')

// Ours
const { freqOf, minOf, maxOf, round, varOf } = require('./.internal/stats')
const {
  cleanTrace,
  extractFrames,
  dumpTree,
  sortEvents,
  toSec
} = require('./.internal/trace-util')

const { sqrt } = Math

const analyzeByCategories = model =>
  dumpTree(model.bottomUpGroupBy('Category'), 'totalTime')

const analyzeByEvents = model =>
  dumpTree(model.bottomUpGroupBy('EventName'), 'totalTime')

const analyzeFunctions = model =>
  dumpTree(model.bottomUp(), 'selfTime')

const analyzeFirstPaint = events => {
  const firstPaint = events.find(e => e.name === 'firstPaint')
  if (!firstPaint) return null

  const first = events.find(e => e.name === 'TracingStartedInPage')
  return toSec(firstPaint.ts - first.ts)
}

const analyzeFps = frames => {
  if (frames.length === 0) return null

  frames = frames.map(f => toSec(f.ts))
  const fpsPerFrames = frames.reduce((fpsPerFrames, f, i, frames) => {
    if (i > 0) {
      const fps = 1 / (f - frames[i - 1])
      fpsPerFrames.push(fps)
    }
    return fpsPerFrames
  }, [])

  // Compute min
  const lo = round(minOf(fpsPerFrames), 2)
  // Compute max
  const hi = round(maxOf(fpsPerFrames), 2)
  // Compute mean
  const mean = round(freqOf(frames), 2)
  // Compute variance
  const variance = round(varOf(fpsPerFrames, mean), 2)
  // Compute standard deviation
  const sd = round(sqrt(variance), 2)

  return { mean, variance, sd, lo, hi }
}

const createReport = (file, events) =>
  new Promise(resolve => {
    events = sortEvents(cleanTrace(events))
    const model = new DevtoolsTimelineModel(events)

    const report = {
      name: path.basename(file),
      profiling: {
        categories: analyzeByCategories(model),
        events: analyzeByEvents(model),
        functions: analyzeFunctions(model)
      },
      firstPaint: analyzeFirstPaint(events),
      fps: analyzeFps(extractFrames(events))
    }
    resolve(report)
  })

module.exports = createReport
