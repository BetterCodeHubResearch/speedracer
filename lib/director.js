// Packages
const debug = require('debug')('director')

// Ours
const Run = require('./run')

class Director {
  constructor(runner, driver, reporter) {
    this.runner = runner
    this.driver = driver
    this.reporter = reporter
    this.file = ''
    this.run = null
    this.runs = []

    this.reset()
  }

  reset() {
    const ack = type => () => this.runner.acknowledge(type)

    this.driver.removeAllListeners()
    this.runner.removeAllListeners()

    this.driver.on('status', status =>
      this.reporter.updateStatus(status)
    )

    this.runner.on('run:start', infos => {
      this.run = new Run(this.file, infos)
      this.reporter.startRun(this.run)
      this.driver.startTracing()
        .then(ack('run:start'))
    })

    this.runner.on('run:end', () => {
      this.driver.stopTracing()
        .then(events => {
          debug('run ended %s (%d events)', this.run.title, events.length)
          this.run.attachEvents(events)
          this.runs.push(this.run)
          this.reporter.finishRun(this.run)
          this.run = null
        })
        .then(ack('run:end'))
    })
  }

  runFile(file) {
    return new Promise(resolve => {
      // no runs timeout
      const timerId = setTimeout(() => {
        debug('no runs found: %s', file)
        this.reporter.warn(`No runs found in ${file}.`)
        this.reset()
        resolve(this.runs)
      }, 1000)

      this.driver.loadFile(file).then(() => {
        debug('file loaded: %s', file)
        this.reporter.startFile(file)
        this.file = file
      })

      this.runner.once('run:start', () => {
        debug('clear timeout')
        clearTimeout(timerId)
      })

      this.runner.once('file:end', () => {
        debug('total runs: %d', this.runs.length)
        this.reporter.finishFile(file)
        this.reset()
        resolve(this.runs)
      })
    })
  }

  close() {}
}

module.exports = modules => {
  return new Director(modules.runner, modules.driver, modules.reporter)
}
