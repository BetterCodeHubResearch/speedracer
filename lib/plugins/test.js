const path = require('path')

const { createDir } = require('../.internal/util')

export default (options) => (reports, file) => {
  reports.forEach(report => {
    const dirname = path.join(path.dirname(file), options.snapshotsDirname)
    const filename = path.join(dirname, `${report.metadata.id}.speedracer`)

    return loadReport(filename).then(
      oldReport => {
        process.exit(0)
      },
      () => createDir(dirname).then(
        () => writeFile(filename, race.report.serialize())
      )
    )
  })
}
