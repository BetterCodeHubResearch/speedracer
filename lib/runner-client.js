class RunnerClient {
  constructor() {
    this.port = 3001
    this.ws = new WebSocket(`ws://localhost:${this.port}`)
    this.router = {
      'run:start:ack': () => this.startRun(),
      'run:end:ack': () => this.runNext()
    }
    this.runs = []

    this.ws.onopen = () => {
      this.ws.onmessage = ({ data }) => this.onMessage(JSON.parse(data))
      this.send('file:start')
      this.runNext()
    }
  }

  enqueueRun(title, fn) {
    this.runs.push({ title, fn })
  }

  runNext() {
    this.currentRun = this.runs.shift()

    if (this.currentRun) {
      this.send('run:start', this.currentRun)
    }
    else {
      this.send('file:end')
      this.ws.close()
    }
  }

  startRun() {
    try {
      const res = this.currentRun.fn()
      if (res && res.then) {
        res.then(() => this.endRun())
      }
      else {
        this.endRun()
      }
    }
    catch (err) {
      this.runError(err)
    }
  }

  endRun() {
    this.send('run:end', this.currentRun)
  }

  runError(err) {
    this.send('run:error', this.currentRun, err)
    this.endRun()
  }

  onMessage(payload) {
    const { type } = payload
    const handler = this.router[type]
    if (handler) {
      handler()
    }
  }

  send(type, run, err) {
    const payload = { type }
    if (run) payload.run = run
    if (err) payload.err = err
    this.ws.send(JSON.stringify(payload))
  }
}

const runner = new RunnerClient()

module.exports = (title, fn) =>
runner.enqueueRun(title, fn)

// For test purposes
module.exports.RunnerClient = RunnerClient
