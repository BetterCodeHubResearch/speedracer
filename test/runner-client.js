// Packages
import sinon from 'sinon'
import test from 'ava'

// Ours
import './helpers/mock-ws'
import { RunnerClient } from '../lib/runner-client'

const CI = !!process.env.CI

const checkMessageSequence = (t, client) => {
  t.snapshot(client.ws.send.args)
}

test.cb('run a race', t => {
  const fn = sinon.spy()
  const client = new RunnerClient()
  client.enqueueRace('foo', fn)
  setTimeout(() => {
    t.true(fn.calledOnce)
    checkMessageSequence(t, client)
    t.end()
  }, CI ? 100 : 0)
})

test.cb('run an async race', t => {
  const fn = sinon.stub().callsFake(() => Promise.resolve())
  const client = new RunnerClient()
  client.enqueueRace('foo', fn)
  // Use setTimeout here because of Promise#then
  setTimeout(() => {
    t.true(fn.calledOnce)
    checkMessageSequence(t, client)
    t.end()
  }, CI ? 100 : 10)
})

test.cb('run multiple races in band', t => {
  const client = new RunnerClient()
  const fn1 = sinon.spy()
  const fn2 = sinon.stub().callsFake(() => Promise.resolve())
  const fn3 = sinon.spy()
  client.enqueueRace('foo', fn1)
  client.enqueueRace('bar', fn2)
  client.enqueueRace('baz', fn3)
  // Use setTimeout here because of Promise#then
  setTimeout(() => {
    t.true(fn1.calledOnce)
    t.true(fn2.calledAfter(fn1))
    t.true(fn3.calledAfter(fn2))
    checkMessageSequence(t, client)
    t.end()
  }, CI ? 100 : 10)
})

test.cb('handle exception in a race', t => {
  const client = new RunnerClient()
  const fnThrow = () => { throw new Error('error') }
  client.enqueueRace('foo', fnThrow, false)
  client.enqueueRace('bar', fnThrow, true)
  client.enqueueRace('baz', () => {}, false)
  setTimeout(() => {
    checkMessageSequence(t, client)
    t.end()
  }, CI ? 100 : 0)
})
