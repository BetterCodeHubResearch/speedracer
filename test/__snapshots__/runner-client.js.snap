// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute a race 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:finish\\"}",
  ],
]
`;

exports[`execute a run 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:end\\"}",
  ],
]
`;

exports[`execute an async race 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:finish\\"}",
  ],
]
`;

exports[`execute an async run 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:end\\"}",
  ],
]
`;

exports[`execute multiple races in band 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:finish\\"}",
  ],
]
`;

exports[`execute multiple runs in band 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:end\\"}",
  ],
]
`;

exports[`handle exception in a race 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:error\\",\\"race\\":{\\"title\\":\\"foo\\"},\\"err\\":{}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:error\\",\\"race\\":{\\"title\\":\\"bar\\"},\\"err\\":{}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:start\\",\\"race\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"race:finish\\",\\"race\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:finish\\"}",
  ],
]
`;

exports[`handle exception in a run 1`] = `
Array [
  Array [
    "{\\"type\\":\\"file:start\\"}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:error\\",\\"run\\":{\\"title\\":\\"foo\\"},\\"err\\":{}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"foo\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:error\\",\\"run\\":{\\"title\\":\\"bar\\"},\\"err\\":{}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"bar\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:start\\",\\"run\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"run:end\\",\\"run\\":{\\"title\\":\\"baz\\"}}",
  ],
  Array [
    "{\\"type\\":\\"file:end\\"}",
  ],
]
`;
